### ISSUE TITLE
Do not call multiple time the same dependency target when serving multiple app in parallel

### ISSUE THREAD

**Issue by jogelin** [nrwl/nx#9021]
- [ ? ] I'd be willing to implement this feature ([contributing guide](https://github.com/nrwl/nx/blob/master/CONTRIBUTING.md))

## Description
In micro frontend architecture, it is usual to start multiple applications in parallel by using the `@nrwl/workspace:run-commands` executor. 

However, when multiple apps are sharing the same dependency target, they both call and run it at the same time.

This is not only in mfe architecture but any time you are running multiple apps in parallel.

## How to reproduce it?

1. Create a new mfe workspace with the Angular presets:
```
npx create-nx-workspace ng-mfe --packageManager=yarn
```

2. Create a new shared library:
```
nx g @nrwl/angular:lib shared/api-models`
```

3. Create a dependency from  `dashboard` to `shared-api-models`. You can create an interface in `shared-api-models` and use it, import it in `dashboard`.

4. Create a dependency from  `login` to `shared-api-models` by doing the same as the bullet above. So you'll will have a dependency graph that looks like:
<img width="465" alt="image" src="https://user-images.githubusercontent.com/954509/154640618-9f33635d-f11a-46aa-99f0-b3573eef3761.png">

5. Then add a new target `generate-models` in `shared-api-models` configs:
```json
    "shared-api-models": {
      "architect": {
        "generate-models": {
          "builder": "@nrwl/workspace:run-commands",
          "options": {
            "commands": ["echo 'REMOVE AND GENERATE API MODELS'"]
          }
        }
      }
    }
```

6. In the `nx.json`, create a global dependency on the `generate-models` target when serving any app:
```json
  "targetDependencies": {
    "serve": [
      {
        "target": "generate-models",
        "projects": "dependencies"
      }
    ]
  },
```

7. Run the target `serve-mfe` on the `dashboard` to serve both the `dashboard` application and the `login` application:
```bash
yarn nx run dashboard:serve-mfe
```

**You'll finally see that the target `generate-models` is executed twice even if it's the same target on the same library:**

![image](https://user-images.githubusercontent.com/954509/154644911-d3a121cf-a9e0-480f-8be5-b3a17d3a59fd.png)

## Motivation
As you can see in the example, I am using this approach to generate APIs models from an openapi and it doesn't make sense to generate them multiple times. It can even create conflicts when the target is modifying/generating files at the same time.

IMO this is the same for a `build` target, it doesn't make sense to build twice the library.

## Suggested Implementation
I understand that each `serve` target are run in parallel so it could be hard to know that another task will execute the same target. 

## Alternate Implementations
I think that the executor `@nrwl/workspace:run-commands` should be smarter and understand what targets will be executed before we execute them. In this way, a filter can be done.

Maybe another executor that is aware of what is a target and is able to consolidate the target to run and not just execute the commands...

**Comment by FrozenPandaz** [comment]
What about `nx run-many --target serve --projects login,dashboard --parallel`? That will only run `generate-models` once.

**Comment by jogelin** [comment]
Correct! 

Thank you very much @FrozenPandaz 👍
